pkg/virt-handler/hotplug-disk/mount_test.go

```diff
@@ -529,72 +380,6 @@ var _ = Describe("HotplugVolume block devices", func() {
 		Expect(res).To(Equal(testFile))
 	})
 
-	It("Should remove the block device and permissions on unmount", func() {
```

Comment:
As was discussed above in this PR:
- These scenarios are being tested in functional tests (`hotplug.go`).
- Unit testing an external library is not really our responsibility, instead we should functionally test a whole flow using our dependencies to ensure the behavior is the expected one, which is what's happening now.
- Unit testing, in this specific case, is not really possible since we will fail on providing `runc` paths that it thinks are invalid. Currently (without the work of this PR) we have so much duplicated code from `runc` in our code-base that it was manipulated to not fail on such fake paths. Of course duplicating massive amounts of code and manipulating it causes much more damage than good.

I honestly think that we deeply test all these scenarios functionally and this is the way to go.

Reply:
> These scenarios are being tested in functional tests (hotplug.go).


My concern here is that the functests just validate the final result of the operation. I.e. whether volume was added or removed without errors. AFAIK the functests do not check e.g. if the cleanup of the block device and permissions was perfomred after unplug (which is what this test case does).


> Unit testing an external library is not really our responsibility, instead we should functionally test a whole flow using our dependencies to ensure the behavior is the expected one, which is what's happening now.


This exact test case is not a test of an external library. This is a a test of KubeVirt cleanup logic. If its removed from here I suppose that this part should be covered somewhere else then. Otherwise the tests coverage gets decreased.

Reply:
I understand your concern, but as unfortunately the only way to test `runc` without is using a real container (with cgroups subsystems at a known absolute path) + calling virt-chroot, and that's not a unit test anymore but a functional test.

Furthermore, we actually don't test the logic of the libraries we use through unit testing. For example, let's say I'm using a new known library for compressing / decompressing files - do I need to unit test it? The cleanup logic of such library could be critical (if, for example, it leaves a lot of files open and never closes them, or other forms of memory leak) but still I don't see myself unit testing it.

And another last point - this test checks implementation, not behavior. It checks that `b 581:50 rwm` exists in `devices.deny`. What if the implementation changes one day and a slightly different permission is used? What about cgroups v2? how can we even check something like that with eBPFs (only way is decompiling bytecode..)? Eventually we care about the end result ("validate the final result of the operation" as you said), not about the concrete implementation of complicated tasks that are outside of our domain.

The only thing I can think of is a functional test that somehow validates that after removing hotplugs everything is clean and back to normal, but again, validating it is hard (especially for v2) and needed to be done without diving too deep into implementation details.

Reply:
Just one last thing

> Otherwise the tests coverage gets decreased

As far as I understand test coverage metrics only refer to the testing of our code, not including any dependencies. Before this PR a lot of `runc` code was duplicated into our codebase, so it was considered as coverage to test it, but this it wrong to begin with. Now this code is not on our hands and therefore testing it or not does not affect coverage.

Reply:
Removal of the block device is not what runc does actually. This is a part of the KubeVirt cleanup logic. Usually such things are tested by mocking the external dependencies (runc and the OS function to remove the block device) and by setting the expectations that the appropriate calls are made with the correct parameters.

Reply:
Okay I've pushed some new commits that mock runc completely (by generating a mocked cgroupManager) and only expect certain rules to be applied. This way we don't have to be coupled to implementation details.

WDYT? :)